src/storm-dft/builder/DFTBuilder.cpp

```diff
@@ -117,237 +108,331 @@ storm::dft::storage::DFT<ValueType> DFTBuilder<ValueType>::build() {
+    STORM_LOG_ASSERT(mNextId == mElements.size(), "Current id " << mNextId << " and number of elements " << mElements.size() << " do not match.");
+    element->setId(mNextId++);
+    mElements[element->name()] = element;
+}
 
-            for (DFTElementPointer const& child : gate->children()) {
-                newrnk = computeRank(child);
-                if (newrnk > maxrnk) {
-                    maxrnk = newrnk;
-                }
-            }
-            elem->setRank(maxrnk + 1);
-        }
-    }
+template<typename ValueType>
+void DFTBuilder<ValueType>::addBasicElementConst(std::string const& name, bool failed) {
+    addElement(std::make_shared<storm::dft::storage::elements::BEConst<ValueType>>(0, name, failed));
+}
 
-    return elem->rank();
+template<typename ValueType>
+void DFTBuilder<ValueType>::addBasicElementProbability(std::string const& name, ValueType probability, ValueType dormancyFactor) {
+    // Handle special cases
+    if (storm::utility::isZero<ValueType>(probability)) {
+        addBasicElementConst(name, false);
+    } else if (storm::utility::isOne<ValueType>(probability)) {
+        addBasicElementConst(name, true);
+    } else {
+        // TODO: check 0 <= dormancyFactor <= 1
+        addElement(std::make_shared<storm::dft::storage::elements::BEProbability<ValueType>>(0, name, probability, dormancyFactor));
+    }
 }
 
 template<typename ValueType>
-bool DFTBuilder<ValueType>::addRestriction(std::string const& name, std::vector<std::string> const& children,
-                                           storm::dft::storage::elements::DFTElementType tp) {
-    if (children.size() <= 1) {
-        STORM_LOG_ERROR("Restrictions require at least two children");
+void DFTBuilder<ValueType>::addBasicElementExponential(std::string const& name, ValueType rate, ValueType dormancyFactor, bool transient) {
+    // Handle special cases
+    if (storm::utility::isZero<ValueType>(rate)) {
+        addBasicElementConst(name, false);
+    } else {
+        // TODO: check 0 <= dormancyFactor <= 1
+        addElement(std::make_shared<storm::dft::storage::elements::BEExponential<ValueType>>(0, name, rate, dormancyFactor, transient));
     }
-    if (nameInUse(name)) {
-        STORM_LOG_ERROR("Element with name '" << name << "' already exists.");
-        return false;
+}
+
+template<typename ValueType>
+void DFTBuilder<ValueType>::addBasicElementErlang(std::string const& name, ValueType rate, unsigned phases, ValueType dormancyFactor) {
+    // Handle special cases
+    if (storm::utility::isZero<ValueType>(rate)) {
+        addBasicElementConst(name, false);
+    } else if (phases == 1) {
+        // shape=1 reduces to exponential distribution
+        addBasicElementExponential(name, rate, dormancyFactor);
+    } else {
+        STORM_LOG_THROW(phases > 0, storm::exceptions::WrongFormatException, "Erlang distribution of BE " << name << " requires a positive number of phases.");
+        // TODO: check 0 <= dormancyFactor <= 1
```

Comment:
Not a bad thing to do that now

Reply:
I will use a `ConstantsComparator` for the checks.